src/Autofac/Util/TypeExtensions.cs

```diff
@@ -142,25 +137,30 @@ public static bool IsDelegate(this Type type)
     /// </summary>
     /// <param name="type">The type to check.</param>
     /// <returns>True if the type is one of the supported enumerable interface types.</returns>
-    public static bool IsGenericEnumerableInterfaceType(this Type type)
+    public static bool IsGenericEnumerableInterfaceType(this Type type, ReflectionCache cache)
```

Comment:
I'm curious if it'd be worth figuring out _why_ we use all these caches.

For example, I see this method, `IsGenericEnumerableInterfaceType`. It... seems like it's doing a lot to just figure out if a thing implements `IEnumerable<T>`.

Am I reading that wrong? I've followed this through and it seems like a lot of work around `type.GetInterfaces()` and some LINQ.

But I get that LINQ is slower than a dictionary lookup, so I see it's used in two spots:

- `AutowiringPropertyInjector`, where it's searching for what properties will be injected. The list of injectable properties gets cached in _a different cache_ so subsequent lookups will skip this call anyway.
- `MetadataFilterAttribute`, where we've hacked in some support for `IEnumerable<T>` and need to pull out the `T` if it's `IEnumerable`. There's a note in that class that explicitly says we don't support `Lazy<T>` or any other relationships in that attribute - it's a hack.

Which is to say, I'm curious if this specific cache actually has value or if the LINQ is, in many cases, "good enough."

I feel like some of the caching has grown up over time, where we have caches that cache other cached data and possibly we don't need all the "layers."

I'm not saying we don't need caching, just that we may be able to simplify the caching system and the way it needs to be cleared/handled if we remove cache layers that aren't actually helping so much anymore.